### DISCUSSION TITLE
Are you happy with helix's keybinding model? [User Opinions]

### DISCUSSION THREAD

**Discussion by txtyash** [helix-editor/helix#1511] [General]
I'm experiencing inconsistency with helix's keybindings. I wish it mostly followed vim's model or kakoune's alone or both available to choose from. For me the keybindings seem not so efficient comparing them straight with kakoune or vim. I experience more mental overhead and inconvenient keystrokes as compared to kakoune or vim but I love having more inbuilt capabilities and features with blazing fast performance.

This is just my opinion and I'm probably inexperienced to think the helix way. 

I want to know what other users think of the keybinding model?

I'm interested to know what other users think of helix's key binding model?

**Comment by pickfire** [comment]
Not all keys are good yet, there are some changes that I would like to do but haven't go to do it yet, like getting visual mode to be more consistent, having goto extend by default and such.

Quite a few things mentioned here can be implemented, I did tried asking in the chat and seemed like people do agree to some of the suggestions mentioned. https://github.com/helix-editor/helix/issues/165 Also, if you have suggestion or faced any specific pain points, that can be discussed separately or in the same place.

**Comment by matoous** [comment]
For what it's worth, I am fine with it as it is as it was easy to switch from nvim to helix, on the other hand I am looking towards experimenting with some of the proposed new keymaps to see if it will suite me better.

**Comment by kirawi** [comment]
(Slightly relevant?) https://github.com/helix-editor/helix/wiki/Keymap-brainstorm

**Comment by bound-variable** [comment]
I would prefer not having to use the `home` and `end` keys - it's awkward with my keyboard, having to press `Fn`. But I'm a noob, so...

**Reply by valpackett** [reply]
Now you can rebind keys..

**Reply by tadashi90** [reply]
Suffering with the same problem.. Didn't want to rebind stuff, I hate to waste time configuring stuff. To not have to install and configure a bunch of stuff is why I'm using Helix =(

**Reply by reinux** [reply]
It's also not great for documentation when everyone's talking in standard key mappings and you're having to do your own thing.

**Comment by sycured** [comment]
I'm a very long time user of vim and recently of neovim
I'm totally disappointed by the keymapping, it'll be good to have an option in Helix config to have vim mapping (keymap = default/vim) so the vim keymapping file already in helix

If someone has a full keymap like vim, I'm ready to try it… I'm going back to vim/neovim due to the keymap and how long is creating the vim keymap for Helix

**Reply by NNBnh** [reply]
If you want Vim keymaps then Helix is just not for you. Helix keymaps design around multiple selections therefore it should be different. Vim-keys compatibility isn't the project goal...

But maybe Helix could be for you if you try to get use to Helix's keymaps because the multiple selections keymaps design is great, it's quite similar to Vim. If you find it difficult to remember then try [this cheat sheet](https://github.com/helix-editor/helix/wiki/Keymap-brainstorm), once you get used to it, you might appreciate it more.

**Comment by bob3000** [comment]
I pretty much appreciate the intention of a modal editor with a consistent concept of a keymap. The Helix keymap actually seems much more consistent to me then the vim keymap.
However I'm looking back on a period of more then ten years of conditioning my muscle memory to the Vim keymap. I'd even take the effort learn a new keymap because I like Helix's design with it's build in features. I can't do a complete switch yet though because I program in many different languages and I heavily rely on some vim plugins the are not optional to me.
For the time being until the functionality of these plugins become available in Helix in core or via plugin I would have to switch between Helix and Vim because of certain languages and that would obviously end up in total mental confusion and would be the opposite of productive. Does that resonate to somebody?
As much much as I appreciate the consistent Helix keymap I'd still find an option to switch back to classic vim keys amazingly helpful. If  a corresponding configuration setting is not that much desired by the community a copy and paste snippet for the configuration file to get the classic Vim key layout could also be an option. Has anybody ever done that yet by any chance?

**Reply by sudormrfbin** [reply]
A vim-esque keymap would be near impossible to add without significant changes to the core system because of one fundamental difference - the selection first, action second in helix borrowed from kakoune.

In vim, for example if you press `d` it waits for the next movement or textobject key to know which part of the text to delete. But in helix there is no waiting around; a `d` immediately deletes whatever text is selected. Every editing primitive is designed around the invariant that it acts on the currently selected text.

**Reply by bob3000** [reply]
Thanks for the explanation. Keep up the good work. You guys are doing great stuff here :clap:

**Comment by David-Else** [comment]
I have only been trying out Helix for a couple of days, but so far I am thinking:

EDIT: I found out more keys!

- Why not add back some familiar vim keys WHEN they don't clash with the Helix ones? I am missing:

`D` delete till end of sentence (vgld)
`Y` yank till end of sentence (vgly)
`C` change till end of sentence (vglc)

Would this be violating some principle?

I am LOVING the `space` and `f`, `b`, `k` etc. This feels so right :)

I am not comfortable with the selecting first approach yet, but I think it is probably better and could get used to it.

I think people not liking the keys is the number one thing that will stop people switching to Helix, so best to make any changes sooner rather than later.

**Reply by the-mikedavis** [reply]
There's a trick I like for selecting until end of line which is possible because of how helix (or kakoune) handles newlines: <kbd>t</kbd><kbd>Enter</kbd> (or <kbd>f</kbd><kbd>Enter</kbd> can also be a good choice if you want to select the newline itself). Any keybind where you get to slap the keyboard is fun :)

IMO the `D`, `Y`, `C` keybinds don't fit well because they bump with the select-then-act paradigm.

**Reply by pickfire** [reply]
For this I usually do `vgld` to delete to end of line and `vghd` to delete to start of line, I am thinking of removing the `v` but that's not done yet.

**Reply by David-Kunz** [reply]
Yes, `D`, `Y` and `C` are the ones I miss most. However, I wouldn't directly introduce those commands in Helix because of its different philosophy (select first, then perform the action) but I would at least provide a _single_ key to select until the end of the line.

**Reply by the-mikedavis** [reply]
You can bind a key to do this like so:

```toml
[keys.normal]
"$" = ["collapse_selection", "select_mode", "goto_line_end", "normal_mode"]
```

**Reply by valpackett** [reply]
There's no need for the `"select_mode", "goto_line_end", "normal_mode"` dance. I just have

```toml
[keys.normal]
"$" = "extend_to_line_end"
```

**Reply by tjmgregory** [reply]
> Enter (or fEnter can also be a good choice if you want to select the newline itself). Any keybind where you get to slap the keyboard is fun :)
This was my favourite new tip in a while!

**Comment by leeola** [comment]
I mostly changed keybinds to my custom Kakoune. I didn't find anything lacking in my Kakoune binds, but Helix had superior(easier) LSP integration for me, and i think is performing better to boot. Plus it's in a language i like so i can contribute.

The thing i'm most happy about is it has been so configurable. I didn't want to change to a new editor mostly. Helix let me keep most of my Kakoune-UX, and i'm really happy with that configuration support

**Reply by sdothum** [reply]
Care to share your Kakoune for Helix key bindings?

**Comment by getreu** [comment]
Is there an key binding set that does not rely on the `[alt]` key as modifier?

The `[alt]` key combinations are used by the _i3-Window Manager_ in my setup.

**Reply by txtyash** [reply]
U should probably use the windows key as mod for your i3 setup. If it is in an uncomfy position then swap it's position with alt with some key remapping tool.

**Reply by getreu** [reply]
This is how I followed your advice:
> U should probably use the windows key as mod for your i3 setup.

In `/etc/i3/config` set

    set $mod Mod4

>  If it is in an uncomfy position then swap it's position with alt with some key remapping tool.

in `/etc/default/keyboard` set:

    XKBOPTIONS="altwin:swap_lalt_lwin"

---

Can one remap the `[alt]` key as modifier in one go to some other key, lets say `[ctrl]` key, `[shift]+[ctrl]` or `[win]` key?

**Reply by getreu** [reply]
> Is there an key binding set that does not rely on the [alt] key as modifier?

[Keymap refactor · Issue #1179 · helix-editor/helix](https://github.com/helix-editor/helix/issues/1179)

**Comment by JigglyBuff-Kun** [comment]
I just started using the project, so far so good. I am coming from neovim / doom-emacs. I like how a lot of the features are built into the editor, unlike vim, where you have to install plugins. Going forward, I hope this is how the project grows, with lots of sane defaults built into the editor rather than having to configure everything. This will bring more people who are put off by configuring vim to use helix. 
Regarding keybindings, I am still getting used to it. I do like the idea of helix's approach, this is the first time I am using it. 

My question to other users who use vim keybindings in other software, how are you guys changing your workflows to adapt to helix ?
For e.g. : I use doom-emacs to use org mode(evil mode), I think it would be a pain to use it after using helix for a while. Also, will helix have something like org mode in the future included in the core ?

**Reply by pickfire** [reply]
>  For e.g. : I use doom-emacs to use org mode(evil mode), I think it would be a pain to use it after using helix for a while. Also, will helix have something like org mode in the future included in the core ?

org-mode I think not anytime near, and it doesn't seemed to be able to be done with LSP I think, so most likely it needs to wait for plugin

**Comment by jficz** [comment]
Using Helix for a few weeks now after using Vim for like 20 years. It's not as bad as I have expected but I still have a few issues:
- `Home` & `End` keys are awkward, I still keep hitting `0` instead of `Home`. I kinda like the idea presented in #2131 (i.e. using `H` & `L`)
- I have trouble with `d` and `x`, mainly because I'm used to Vim's `x` and `dd` which makes my line and char delete operations a pain. But I guess this is more about me not yet being used to the select-first paradigm
- not sure if bug or feature - when `x` is pressed on empty line, it selects the line below it as well. On top of that pressing `d` on empty line deletes the line. This makes deleting whole lines inconsistent
- again, not sure if bug or feature - when I press PageDown and then PageUp I end up on a different line. I've got a ThinkPad with PgUP/Down right next to arrow keys and sometimes I press them accidentally which destroys my flow because I can't easily go back to the place I drifted from
- I really miss `V` and `C-v` features (line select and block select) - I know `V` can somewhat be emulated by `x` but `C-v` does not have an equivalent I could find. The closest thing would be split-line multi-cursor select which is much more complex in comparison
- Also, `ZZ` :)

**Reply by CptPotato** [reply]
> not sure if bug or feature - when x is pressed on empty line, it selects the line below it as well. On top of that pressing d on empty line deletes the line. This makes deleting whole lines inconsistent

I addressed the same issue here: #835

**Reply by sbromberger** [reply]
If it helps, I mapped `ZZ`:
```
Z = { Z = ":wq" }
```
It's not quite the same, but it's good enough for my use.

**Reply by vtmx** [reply]
```
[keys.normal]
Z = { Z = ":wq" }
```

**Comment by sycured** [comment]
Someone did a vim-like config for helix: https://github.com/LGUG2Z/helix-vim

So maybe vim users can migrate more easily to helix… without having to relearn absolutely all keymaps

**Reply by LGUG2Z** [reply]
I noticed a few of the Helix maintainers starred it, though I'm not sure if it's because they like the idea/execution or if it's because they like the `README`. 😅

**Reply by the-mikedavis** [reply]
Admittedly, reading through `config.toml` did put my hair on end a little but overall I'm very impressed how close to vim you can get with some determination. Plus of course the readme is great 😄

We're looking to merge a fix for #2051 soon - maybe before the next release (no promises though) - not that I can ever condone using Helix like this 🙂😛

**Reply by LGUG2Z** [reply]
Just looking through all the requests/demands for a Vim mode in various issues/discussions/PRs here, I have a lot of empathy for all of the contributors and maintainers here.

From my experience writing and maintaining [komorebi](https://github.com/LGUG2Z/komorebi), which aims to be somewhat similar to two incumbent tiling window managers (`bspwm` and `yabai`), but different enough in key areas that justified the creation of the project, I have been met with many users who want it to shed its peculiarities and be _the same as_ those twms (kind of like the Kakoune comments here), or alternatively, they want it to be changed to be like something totally different like `i3` (kind of like the Vim comments here).

I hope to be able to get this configuration to a state where, when people comment asking you to "make this like VIM!!!?!?!?" in the issues and discussions in the future, it's easy to link to and say "Here you go, try this at your own risk" and end those conversations without having to put so much mental energy into the endless back-and-forth justifying the design and vision of this project.

**Comment by diktomat** [comment]
I mostly like it, only added H/M/L keys I'm used to from Vim (see below). What I'm constantly stumble upon though is the handling of newlines as regular characters. `xc` putting me on beginning of the next line instead of keeping the line I'm on, `x` on an empty line selecting that one and the next line, etc. I get that it's more consistent, it just doesn't fit my mind apparently.

```toml
[keys.normal]
H = "goto_window_top"
M = "goto_window_center"
L = "goto_window_bottom"
```

**Reply by ChrHorn** [reply]
> I mostly like it, only added H/M/L keys I'm used to from Vim (see below). What I'm constantly stumble upon though is the handling of newlines as regular characters. `xc` putting me on beginning of the next line instead of keeping the line I'm on,

You can use `x_c` to imitate vim's `cc` behavior. It is a bit clunky, but once https://github.com/helix-editor/helix/pull/2634 is merged it will be possible to rebind it.

**Reply by the-mikedavis** [reply]
> x on an empty line selecting that one and the next line, etc

You may want to use `X` when you think "select the current line" rather than `x`. `x` is inconsistent with the rest of normal mode actually and to have it behave like the other motions, you can remap like so:

```toml
[keys.normal]
x = "extend_to_line_bounds" # previously "X"
X = "no_op"
[keys.select]
x = "extend_line" # previously normal mode "x"
```

**Comment by noor-tg** [comment]
I was using `doom emacs` so I used to do `space space` to open project files list. and I mapped it to hilex config with 
```
[keys.normal]
space = { space = "file_picker" } 
```
I think it is better than using `space f`

**Reply by langston-barrett** [reply]
Similarly, coming from Spacemacs, I use
```
space = { space = "command_palette" } 
```

**Comment by workingj** [comment]
i'am pretty much happy with the std key mapping of helix, find it very intuitive.
But for me learning and relearning keys is also not so much an issue.
So i go with the build in, except a few additional commands mapped for my work flow.

**Comment by jharrilim** [comment]
I don't really have any gripes with any individual bindings. I can say that the thing I enjoy the most about helix is that it opens modal dialogs for its different modes (modal modals?), and makes it intuitive to discover/rediscover keybindings. The command palette is super useful too, and is a great way to discover what some of the keybindings are.

There is a nifty trick you can use from the command palette (`space`, `?`) to search for keybindings: Type in `(`, some key, and then `)` to "search by binding". ie. `(C)` will show you that the `C` binding is "Copy selection on next line"

All of the features around discoverability and useability is what wins for me. It takes a lot of weight off of individual key bindings. In comparison to vim, well, try doing a Google search for "vim keybindings" and "helix keybindings". vim has `:help tutor`, but helix has `:tutor`, [docs.helix-editor.com/keymap.html](https://docs.helix-editor.com/keymap.html), and all of the dialogs you get while editing.

I think as long as good discoverability remains a _key_ theme for helix, I think the weight on individual key choice selection matters a lot less. Not to say that it doesn't matter entirely of course, but I think I'll be happy with most keybindings that are given to me, so long as there is documentation and that they're searchable.

**Comment by Nathan-Furnal** [comment]
I'm an Emacs user and being able to use `Ctrl` and `Alt` instead of the modal experience would be amazing.

**Reply by CptPotato** [reply]
I think this is at odds with what helix aims to be.

You can however customize your insert mode key map and introduce <kbd>Ctrl</kbd> / <kbd>Alt</kbd> bindings to your liking.

**Reply by Nathan-Furnal** [reply]
Thanks, it was worth a try :smile:

**Reply by pycaw** [reply]
Same. Why is there no bindings off the shelf for "common keybindings", that is, ctrl+c/v to copy/paste, etc?

**Reply by archseer** [reply]
Because we follow vi/kakoune style bindings? y/p, etc

**Reply by pycaw** [reply]
> Because we follow vi/kakoune style bindings? y/p, etc

But why not accomodate this, that is, the non-modal style too? Bindigs are flexible already, only some additional code would be required to have this and that flex a bit more. 
Think about how much effort game devs go through to support mouse, controllers, VR, etc. Here, it's surely simpler.

**Reply by archseer** [reply]
Because the bindings are really flexible you're able to bind them yourself if you'd like the editor to behave differently than it was designed for

**Reply by pycaw** [reply]
> Because the bindings are really flexible you're able to bind them yourself if you'd like the editor to behave differently than it was designed for

You are not helping

**Reply by haxfn** [reply]
@pycaw please read "Modal editing"

> > Because we follow vi/kakoune style bindings? y/p, etc
> 
> But why not accomodate this

Because modal editor have to be modal. That is what modal editors do. Accommodating one non-modal thing justifies accommodating every non-modal thing.

Why use a modal editor if you don't want leave Ctrl+c. Try unlearning: use "Space+p" or "Space+y" for paste or yanking(copying). Stick with your usual WYSIWYG editor if you don't want to unlearn.

By the way depending on your terminal Ctrl+v would still work as expected if you are in Insert mode.

**Reply by haxfn** [reply]
> You are not helping

Your comment was unhelpful itself instead. What did you not find helpful? What did you not understand, convey that if you need help.

What is your motivation to switching to helix? What should you do to copy/paste from the system clipboard? Convey/Ask all that. Don't say, "You are not helping" instead say "I still don't understand this"

**Reply by pycaw** [reply]
> > You are not helping
> 
> Your comment was unhelpful itself instead. What did you not find helpful? What did you not understand, convey that if you need help.
> 
> What is your motivation to switching to helix? What should you do to copy/paste from the system clipboard? Convey/Ask all that. Don't say, "You are not helping" instead say "I still don't understand this"

My motivation is to use a modern, decent terminal-based editor with LSP support and no modal UI forced down my throat. I use micro but LSP support is lacking. Emacs doesn't classify as modern and is too quirky (tried it as well and ditched it long ago). So Helix would be great if it extended its UI to work out-of-the-box with non-modal style. I myself, naively, don't see why that is such an impossibility or against its core design principles.

**Reply by haxfn** [reply]
> no modal UI forced down my throat.

Don't try swallowing modal editing either try learning modal editing first or try @pycaw try this maybe [microsoft/edit](https://github.com/microsoft/edit)

> be great if it extended its UI to work out-of-the-box with non-modal style. I myself, naively, don't see why that is such an impossibility or against its core design principles.

Maybe someone else will do a better job at explaining that to you.

**Reply by jficz** [reply]
@pycaw 
> My motivation is to use a modern, decent terminal-based editor with LSP support and no modal UI forced down my throat. 

Then Helix is just not for you. Modal editing is a design principle editors like Helix and Vim are based on. If you don't want modal, you don't want Helix, simple as that.

**Reply by pycaw** [reply]
> @pycaw
> 
> > My motivation is to use a modern, decent terminal-based editor with LSP support and no modal UI forced down my throat.
> 
> Then Helix is just not for you. Modal editing is a design principle editors like Helix and Vim are based on. If you don't want modal, you don't want Helix, simple as that.

Sad to hear that. Hope someone will soon fill the gaping void.

**Reply by heyakyra** [reply]
@pycaw [on Jun 26](https://github.com/helix-editor/helix/discussions/1511#discussioncomment-13584064):
> Sad to hear that. 

Why is that sad? It's like going to a burger restaurant and being heartbroken there is no chinese buffet. 

@pycaw on Jun 21:
> My motivation is to use a modern, decent terminal-based editor with LSP support and no modal UI forced down my throat. I use micro but LSP support is lacking. Emacs doesn't classify as modern and is too quirky (tried it as well and ditched it long ago). So Helix would be great if it extended its UI to work out-of-the-box with non-modal style. I myself, naively, don't see why that is such an impossibility or against its core design principles.

You and @Nathan-Furnal probably want:

https://github.com/zee-editor/zee

There's also Lapce, but it is modal and has a gui which was discussed here (as it sadly for the rest of us has no Kakoune/Helix-style key bindings…yet):

https://github.com/helix-editor/helix/discussions/11783#discussioncomment-11040502

**Reply by pycaw** [reply]
> @pycaw [on Jun 26](https://github.com/helix-editor/helix/discussions/1511#discussioncomment-13584064):
> 
> > Sad to hear that.
> 
> Why is that sad? It's like going to a burger restaurant and being heartbroken there is no chinese buffet.
> 
> @pycaw on Jun 21:
> 
> > My motivation is to use a modern, decent terminal-based editor with LSP support and no modal UI forced down my throat. I use micro but LSP support is lacking. Emacs doesn't classify as modern and is too quirky (tried it as well and ditched it long ago). So Helix would be great if it extended its UI to work out-of-the-box with non-modal style. I myself, naively, don't see why that is such an impossibility or against its core design principles.
> 
> You and @Nathan-Furnal probably want:
> 
> https://github.com/zee-editor/zee
> 
> There's also Lapce, but it is modal and has a gui which was discussed here (as it sadly for the rest of us has no Kakoune/Helix-style key bindings…yet):
> 
> [#11783 (comment)](https://github.com/helix-editor/helix/discussions/11783#discussioncomment-11040502)

No, I'd say it's more like wanting to a play a simulator on PC with a joystick as opposed to an xbox controller. And there is no subtitute available, hence demand is that much more legitimate.

**Comment by Sepero** [comment]


**Reply by gabydd** [reply]
You can rebind ``C-w``p, for example to make it do nothing in insert mode
```
[keys.insert]
 C-w = "no_op"
```

Currently you cant rebind a group https://github.com/helix-editor/helix/issues/1508 so you would have to manually add all the keybindings for window mode if you wanted C-w to have those bindings

**Reply by CaesarXInsanium** [reply]
I like the insert mode C-w, I can practice my spelling by deleting the entire word.
Also why would anyone leave the window in the middle of writing

**Comment by Sepero** [comment]


**Reply by pascalkuthe** [reply]
All of these insertmode bindings that go to normal mode and undo/save and go back to insert mode won't work and will cause crashes. Insert mode bindings aren't really supported by helix. Insert mode is intended only for inserting text, not any additional actions

**Comment by David-Else** [comment]
:monocle_face:
```
"A-s" = [ ":set soft-wrap.enable true" ]
"A-S" = [ ":set soft-wrap.enable false" ]
```
:beers:
```
[keys.normal.space]
t = { s = ":toggle soft-wrap.enable" }
```
PS You might need the latest Helix version.

**Comment by TeddyDD** [comment]
I tried to use Helix, but the key map is the reason I decided against it. Helix uses bastardized version of Kakoune bindings. Kakoune differs from Vim, but at least there is the [proof](https://github.com/mawww/golf/) changes are for the better. Vim bindings have the advantage of familiarity. There is no merit behind Helix bindings (I'm not counting wave handy „not want to press modifiers”), just [design by committee](https://github.com/helix-editor/helix/issues/165).

**Reply by kirawi** [reply]
Helix's keymap is designed to be memorable and easy to use, and we do take possible improvements into consideration. Number of strokes are not the goal, as per our vision: https://github.com/helix-editor/helix/blob/master/docs/vision.md

It would be more helpful to discuss where Helix's keymap falls short for you. We can also discuss what "better" means for you.

**Reply by archseer** [reply]
Ah yes, code golf: the ultimate proof of editor efficiency.

**Reply by archseer** [reply]
There's no design by committee. I've designed the keybinds the way I preferred them (and they mostly match kakoune, I don't find vww any less efficient than wW since you're still pressing the same amount of keys. Shift counts too)

If you don't like the mappings we of course allow key remapping, but sticking with a different editor is a perfectly fine alternative too.

**Reply by TeddyDD** [reply]
> Ah yes, code golf: the ultimate proof of editor efficiency.

Code golf might not be the best metric, but at least it is an objective metric you can use and compare keymap with vim. What I'm trying to say is Vim bindings have huge value of being ubiquitous, and to change them you should have good reasons.

> Number of strokes are not the goal, as per our vision

Fair enough

> vww any less efficient than wW since you're still pressing the same amount of keys. Shift counts too

It's less efficient since you are dealing with one additional mode. The mental overhead is huge. I suspect that's one of the reasons why Vim golf counts modifier + key as a single keystroke.

At this point Helix kemap is like vim's with visual mode and multi cursors plugin + some visual feedback. Minor consistencty fixes won't change that.

> sticking with a different editor is a perfectly fine alternative too.

Sure, sticking with your keybindings is fine as well. I'm not a target audience for Helix. Just wanted to express my opinion.

**Reply by dead10ck** [reply]
> Code golf might not be the best metric, but at least it is an objective metric you can use and compare keymap with vim. What I'm trying to say is Vim bindings have huge value of being ubiquitous, and to change them you should have good reasons.

There is nothing objective about personal preferences. There is nothing inherently better about objectivity when comparing a subjective experience. This is like saying saying one blanket is better than another because its thread count is higher.

> Sure, sticking with your keybindings is fine as well. I'm not a target audience for Helix. Just wanted to express my opinion.

Expressing constructive feedback is always welcome, even if you decide helix is not for you. Just showing up to say "lol vim golf 🧐" is just trolling.

**Comment by EnricoGhiorzi** [comment]
I use Helix as my first modal editor (tried NeoVim and Kakoune before but failed to stick to them). I now use it as my main editor for pretty much everything. This alone says a lot about Helix's user friendliness! Though, as a user, there is something I cannot wrap my head around. To use a command, I can either:

- Use a keybind, if I can remember it
- Press ':' and get into command mode, which uses a discover-as-you-type model
- Press space and get into space mode from where I have a tree-menu-like with some commands
- Into space mode, open the command palette with fuzzy-finding

Commands can be accessible in any of these ways without an apparent consistent structure: some are duplicated in multiple modes, others are accessible in only one way. Only command-mode commands can apparently take arguments. This is confusing and I wish there was a more unified model.

**Reply by kirawi** [reply]
I want to provide some context:
- Helix began without any keymap customization for a while until it was tacked on as a band-aid solution until the plugin system could supersede it (https://github.com/helix-editor/helix/pull/268). The possibility of unifying typable and regular commands was also acknowledged, thought it seems it was lost to history (https://github.com/helix-editor/helix/pull/268#pullrequestreview-684658757).
- https://github.com/helix-editor/helix/pull/454 added support for chained keybindings
- `space-?` was added for command discoverability which became more useful once we started adding commands without assigned keys (https://github.com/helix-editor/helix/issues/559). Prior to this, the recommendation was looking at the source code.

**Comment by thiagomajesk** [comment]
I posted a comment just recently about my experience coming back to Helix with a different perspective… Just gonna link my previous two comments here for visibility (and future reference): 1) https://github.com/helix-editor/helix/issues/165#issuecomment-1771544419 and 2) https://github.com/helix-editor/helix/issues/165#issuecomment-1771692381.

**Comment by goyalyashpal** [comment]
helix just didn't work for me. there have been constant bugs and bugs and bugs in my experience with it.

just on the distribution side:

1. _this one has been long fixed, so not mentioning it_ 
2. then this `msys2-pacman + hx = 🐛`: https://github.com/helix-editor/helix/discussions/8017
3. this `wezterm + hx = 🐛` one still persists: https://github.com/wez/wezterm/issues/5213 
which when fixed would replace this one: https://github.com/helix-editor/helix/issues/3162

after sidelining these distribution problems, even the usability part of helix has been horrible for me unfortunately:
1. first major one was this `👀` : https://github.com/helix-editor/helix/discussions/5794
2. and this `🔲⬆️`  : https://github.com/helix-editor/helix/discussions/5548#discussioncomment-4694127
3. and the biggest of all, the poor poor line `⬌` selection: https://github.com/helix-editor/helix/discussions/5764#discussioncomment-4854530
4. then this one finally made me give up on helix for good `⧉⌧🔳`: https://github.com/helix-editor/helix/issues/6435

not to mention the 5-7 others which i was willing to put up with. and many many of the following recurring theme:

> _per1:_ no good use case, no point in making this new feature configurable; 
> _per2:_ but this breaks my work & is doable in neo/vim

i have just re-continued my learning journey on `neovim`, and it's ohkay so far too.
yes, helix was kinda my first dedicated venture into modal editors

**Comment by haxfn** [comment]
I had used neovim for about 3 years then four months ago, when configuring rust was a pain I decided to use helix cold-turkey, never looked back at neovim.

Sufficiently consistent keybindings when editing and sufficiently synergetic modal editing, Discoverability of features, ready-to-use, fast (pasting huge files in a zap).

Rough edge: inconsistent keybindings. file-pickers not using same keybindings, command mode ':' not using same keybindings.

Rough edge: hackability. In terms of laying bare components of the editor (project name, file editing time, list of buffers open) and writing my own cute programs with them for keybindings.

What do others think about my comments?

**Comment by azymohliad** [comment]
I like most of the defaults. And I definitely love how discoverable things are when configuring custom bindings: `<space>?` to fuzzy-find the exact action name, `%b` filter to find action by binding or check if the binding is already assigned.

My main problem with the defaults is with pairs of complementary actions assigned to far-apart keys by acronym: `n`/`p` (for "next" and "previous" when used with modifiers), `u`/`d` (for "up" and "down" when used with modifiers). I find it very inconvenient, and I would deviate even further from Vim to avoid this. In other places Helix uses lower/upper case (`u`/`U`, `p`/`P`, `o`/`O`, `n`/`N`, etc) and adjacent keys (`j`/`k`, `[`/`]`, `<`/`>`) for complementary action pairs, and I think both of these approaches are much better.

I also find the default multi-cursor expansion with `C` and `A-C` a bit inconsistent and unintuitive. I understand that `c` was already taken, so it couldn't use the lower/upper case pair and went for `A-` modifier instead, but it's very different from other action pairs with opposite directions. I changed it to `A-J` and `A-K`, and am very happy with it.

But since the defaults can be easily changed, those issues are very minor.

<details>
<summary>My custom bindings so far (admittedly, a bit VSCode-y)</summary>

```toml
[keys.normal]
# Navigation
"C-j" = "page_cursor_half_down"
"C-k" = "page_cursor_half_up"
"C-h" = "goto_line_start"
"C-l" = "goto_line_end"
"A-]" = "select_next_sibling"
"A-[" = "select_prev_sibling"
# Move line up/down
"A-j" = [ "extend_to_line_bounds", "delete_selection", "paste_after" ]
"A-k" = [ "extend_to_line_bounds", "delete_selection", "move_line_up", "paste_before" ]
# Duplicate line up/down
"C-J" = [ "extend_to_line_bounds", "yank", "paste_after" ]
"C-K" = [ "extend_to_line_bounds", "yank", "paste_before" ]
# Add cursor up/down
"A-J" = "copy_selection_on_next_line"
"A-K" = "copy_selection_on_prev_line"
# Discard multi-cursor
"esc" = "keep_primary_selection"
# Select line up (mirror x for selection down)
"X" = "extend_line_above"
# Select a word
"A-w" = [ "move_prev_word_start", "move_next_word_end" ]
"A-W" = [ "move_prev_long_word_start", "move_next_long_word_end" ]
# Comment
"C-/" = "toggle_comments"

[keys.select]
# Skip selection in search (C-k C-d in VScode)
"C-n" = [ "normal_mode", "search_next", "select_mode" ]
```

</details>

**Comment by adiabatic** [comment]
I [made my Helix config](https://github.com/adiabatic/helix-config) available for anyone to look at. Because I think it has interesting ideas in it (as well as some boring ones), I put in a README and a very-WIP explainer.

My main motivator was that I had a lot of trouble getting good at w/e/b for cursor movement. So:

- forward/back by words (broadly defined) are we/sd/xc and their uppercase counterparts (uppercase letters go further)
- jumping around by fractions of a screenful are done by holding Ctrl and pressing er/df/cv
- I had to (re)move stuff (around) to compensate, but this was, surprisingly, not a long slog

**Comment by Kyzzsa** [comment]
The remap below is definitely more intuitive. When pressing `shift`, you get an uppercase word.
``` toml 
"`" = "switch_to_lowercase"
"~" = "switch_to_uppercase"
"A-`" = "switch_case"
```
